File: domains/test/python/src/ingestion/sources.py
"""
Data source definitions for the ingestion module.

This module defines various data sources that can be ingested
by the platform, including files, databases, and APIs.
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from typing import Any, Dict, Iterator, List, Optional, Union
import logging

from core.base import BaseComponent, Metadata, DataFormat
from core.exceptions import DataIngestionError, ConnectionError
from core.config import IngestionConfig


class SourceType(Enum):
    """Types of data sources."""
    FILE = "file"
    DATABASE = "database"
    API = "api"
    STREAM = "stream"
    MEMORY = "memory"


@dataclass
class SourceCredentials:
    """Credentials for data source access."""
    
    username: Optional[str] = None
    password: Optional[str] = None
    api_key: Optional[str] = None
    token: Optional[str] = None
    host: Optional[str] = None
    port: Optional[int] = None
    database: Optional[str] = None
    extra: Dict[str, Any] = None
    
    def __post_init__(self):
        """Initialize extra dictionary."""
        if self.extra is None:
            self.extra = {}
    
    def mask_sensitive(self) -> Dict[str, Any]:
        """Get credentials with sensitive data masked.
        
        Returns:
            Dictionary with masked sensitive values
        """
        masked = {}
        
        if self.username:
            masked['username'] = self.username
        if self.password:
            masked['password'] = '***'
        if self.api_key:
            masked['api_key'] = f"{self.api_key[:4]}...{self.api_key[-4:]}" if len(self.api_key) > 8 else '***'
        if self.token:
            masked['token'] = '***'
        if self.host:
            masked['host'] = self.host
        if self.port:
            masked['port'] = self.port
        if self.database:
            masked['database'] = self.database
        
        return masked


class DataSource(BaseComponent):
    """Abstract base class for data sources."""
    
    def __init__(self, name: str, source_type: SourceType,
                 config: Optional[IngestionConfig] = None,
                 credentials: Optional[SourceCredentials] = None):
        """Initialize data source.
        
        Args:
            name: Source name
            source_type: Type of source
            config: Ingestion configuration
            credentials: Source credentials
        """
        metadata = Metadata(name=name, description=f"{source_type.value} data source")
        super().__init__(metadata)
        
        self.source_type = source_type
        self.config = config or IngestionConfig()
        self.credentials = credentials
        self._connected = False
    
    @abstractmethod
    def connect(self) -> None:
        """Establish connection to the data source.
        
        Raises:
            ConnectionError: If connection fails
        """
        pass
    
    @abstractmethod
    def disconnect(self) -> None:
        """Close connection to the data source."""
        pass
    
    @abstractmethod
    def read(self, **kwargs) -> Iterator[Dict[str, Any]]:
        """Read data from the source.
        
        Args:
            **kwargs: Additional read parameters
            
        Yields:
            Data records as dictionaries
        """
        pass
    
    @abstractmethod
    def get_schema(self) -> Dict[str, Any]:
        """Get schema information for the data source.
        
        Returns:
            Schema dictionary
        """
        pass
    
    def initialize(self) -> None:
        """Initialize the data source."""
        self.connect()
    
    def cleanup(self) -> None:
        """Cleanup data source resources."""
        if self._connected:
            self.disconnect()
    
    def validate(self) -> bool:
        """Validate data source configuration.
        
        Returns:
            True if configuration is valid
        """
        return self.config.validate()
    
    def is_connected(self) -> bool:
        """Check if source is connected.
        
        Returns:
            True if connected
        """
        return self._connected


class FileSource(DataSource):
    """Data source for file-based data."""
    
    def __init__(self, file_path: Union[str, Path], 
                 format: Optional[DataFormat] = None,
                 config: Optional[IngestionConfig] = None):
        """Initialize file source.
        
        Args:
            file_path: Path to file
            format: File format
            config: Ingestion configuration
        """
        self.file_path = Path(file_path)
        
        # Auto-detect format from extension if not provided
        if format is None:
            ext = self.file_path.suffix.lower()[1:]  # Remove the dot
            try:
                format = DataFormat[ext.upper()]
            except KeyError:
                format = DataFormat.TEXT
        
        self.format = format
        
        super().__init__(
            name=f"file:{self.file_path.name}",
            source_type=SourceType.FILE,
            config=config
        )
    
    def connect(self) -> None:
        """Verify file exists and is accessible.
        
        Raises:
            ConnectionError: If file cannot be accessed
        """
        if not self.file_path.exists():
            raise ConnectionError(f"File not found: {self.file_path}")
        
        if not self.file_path.is_file():
            raise ConnectionError(f"Path is not a file: {self.file_path}")
        
        # Check file size
        size_mb = self.file_path.stat().st_size / (1024 * 1024)
        if size_mb > self.config.max_file_size_mb:
            raise ConnectionError(
                f"File size {size_mb:.2f}MB exceeds maximum "
                f"{self.config.max_file_size_mb}MB"
            )
        
        self._connected = True
        self.logger.info(f"Connected to file source: {self.file_path}")
    
    def disconnect(self) -> None:
        """Disconnect from file source."""
        self._connected = False
        self.logger.debug(f"Disconnected from file source: {self.file_path}")
    
    def read(self, **kwargs) -> Iterator[Dict[str, Any]]:
        """Read data from file.
        
        Args:
            **kwargs: Additional read parameters
            
        Yields:
            Data records as dictionaries
        """
        if not self._connected:
            raise DataIngestionError("Source not connected", source=str(self.file_path))
        
        # Import parser for the file format
        from .parsers import get_parser
        
        parser = get_parser(self.format)
        
        with open(self.file_path, 'r', encoding=self.config.encoding) as f:
            for record in parser.parse(f, self.config):
                yield record
    
    def get_schema(self) -> Dict[str, Any]:
        """Get schema for file data.
        
        Returns:
            Schema dictionary
        """
        # Import schema detector
        from .schema import detect_schema
        
        # Read sample of data to detect schema
        sample = []
        for i, record in enumerate(self.read()):
            sample.append(record)
            if i >= 100:  # Sample first 100 records
                break
        
        return detect_schema(sample)


class DatabaseSource(DataSource):
    """Data source for database connections."""
    
    def __init__(self, connection_string: str,
                 query: Optional[str] = None,
                 table: Optional[str] = None,
                 config: Optional[IngestionConfig] = None,
                 credentials: Optional[SourceCredentials] = None):
        """Initialize database source.
        
        Args:
            connection_string: Database connection string
            query: SQL query to execute
            table: Table name to read from
            config: Ingestion configuration
            credentials: Database credentials
        """
        self.connection_string = connection_string
        self.query = query
        self.table = table
        self.connection = None
        
        super().__init__(
            name=f"db:{table or 'query'}",
            source_type=SourceType.DATABASE,
            config=config,
            credentials=credentials
        )
    
    def connect(self) -> None:
        """Connect to database.
        
        Raises:
            ConnectionError: If connection fails
        """
        try:
            # Simplified connection logic
            # In production, would use actual database driver
            self.logger.info(f"Connecting to database: {self.connection_string}")
            
            # Mock connection
            self.connection = {
                'string': self.connection_string,
                'credentials': self.credentials.mask_sensitive() if self.credentials else {}
            }
            
            self._connected = True
            self.logger.info("Database connection established")
            
        except Exception as e:
            raise ConnectionError(
                f"Failed to connect to database: {e}",
                host=self.credentials.host if self.credentials else None
            )
    
    def disconnect(self) -> None:
        """Disconnect from database."""
        if self.connection:
            self.logger.info("Closing database connection")
            self.connection = None
            self._connected = False
    
    def read(self, **kwargs) -> Iterator[Dict[str, Any]]:
        """Read data from database.
        
        Args:
            **kwargs: Additional read parameters
            
        Yields:
            Data records as dictionaries
        """
        if not self._connected:
            raise DataIngestionError("Database not connected")
        
        # Simplified read logic
        # In production, would execute actual SQL query
        
        if self.query:
            self.logger.info(f"Executing query: {self.query[:100]}...")
        elif self.table:
            self.logger.info(f"Reading from table: {self.table}")
        
        # Mock data generation
        for i in range(10):  # Generate 10 mock records
            yield {
                'id': i,
                'value': f"data_{i}",
                'source': 'database'
            }
    
    def get_schema(self) -> Dict[str, Any]:
        """Get database schema.
        
        Returns:
            Schema dictionary
        """
        if not self._connected:
            raise DataIngestionError("Database not connected")
        
        # Mock schema
        return {
            'fields': [
                {'name': 'id', 'type': 'integer', 'nullable': False},
                {'name': 'value', 'type': 'string', 'nullable': True},
                {'name': 'source', 'type': 'string', 'nullable': True}
            ],
            'primary_key': ['id']
        }


class APISource(DataSource):
    """Data source for API endpoints."""
    
    def __init__(self, endpoint: str,
                 method: str = "GET",
                 headers: Optional[Dict[str, str]] = None,
                 params: Optional[Dict[str, Any]] = None,
                 config: Optional[IngestionConfig] = None,
                 credentials: Optional[SourceCredentials] = None):
        """Initialize API source.
        
        Args:
            endpoint: API endpoint URL
            method: HTTP method
            headers: Request headers
            params: Query parameters
            config: Ingestion configuration
            credentials: API credentials
        """
        self.endpoint = endpoint
        self.method = method.upper()
        self.headers = headers or {}
        self.params = params or {}
        
        super().__init__(
            name=f"api:{endpoint}",
            source_type=SourceType.API,
            config=config,
            credentials=credentials
        )
    
    def connect(self) -> None:
        """Establish API connection.
        
        Raises:
            ConnectionError: If connection fails
        """
        # Add authentication headers if credentials provided
        if self.credentials:
            if self.credentials.api_key:
                self.headers['X-API-Key'] = self.credentials.api_key
            elif self.credentials.token:
                self.headers['Authorization'] = f"Bearer {self.credentials.token}"
        
        # Test connection with a simple request
        try:
            self.logger.info(f"Testing API connection to {self.endpoint}")
            # In production, would make actual HTTP request
            self._connected = True
            self.logger.info("API connection established")
        except Exception as e:
            raise ConnectionError(f"Failed to connect to API: {e}", host=self.endpoint)
    
    def disconnect(self) -> None:
        """Disconnect from API."""
        self._connected = False
        self.logger.debug(f"Disconnected from API: {self.endpoint}")
    
    def read(self, **kwargs) -> Iterator[Dict[str, Any]]:
        """Read data from API.
        
        Args:
            **kwargs: Additional read parameters
            
        Yields:
            Data records as dictionaries
        """
        if not self._connected:
            raise DataIngestionError("API not connected", source=self.endpoint)
        
        # Mock API response
        # In production, would make actual HTTP requests with pagination
        
        page = 1
        while page <= 3:  # Mock 3 pages of data
            self.logger.debug(f"Fetching page {page} from API")
            
            # Mock response data
            for i in range(5):  # 5 records per page
                yield {
                    'id': f"api_{page}_{i}",
                    'data': f"value_{page}_{i}",
                    'timestamp': '2024-01-01T00:00:00Z'
                }
            
            page += 1
    
    def get_schema(self) -> Dict[str, Any]:
        """Get API response schema.
        
        Returns:
            Schema dictionary
        """
        # Mock schema based on expected API response
        return {
            'fields': [
                {'name': 'id', 'type': 'string'},
                {'name': 'data', 'type': 'string'},
                {'name': 'timestamp', 'type': 'datetime'}
            ]
        }


class StreamSource(DataSource):
    """Data source for streaming data."""
    
    def __init__(self, stream_url: str,
                 protocol: str = "websocket",
                 config: Optional[IngestionConfig] = None,
                 credentials: Optional[SourceCredentials] = None):
        """Initialize stream source.
        
        Args:
            stream_url: Stream endpoint URL
            protocol: Streaming protocol
            config: Ingestion configuration
            credentials: Stream credentials
        """
        self.stream_url = stream_url
        self.protocol = protocol
        self.stream = None
        
        super().__init__(
            name=f"stream:{stream_url}",
            source_type=SourceType.STREAM,
            config=config,
            credentials=credentials
        )
    
    def connect(self) -> None:
        """Connect to stream.
        
        Raises:
            ConnectionError: If connection fails
        """
        try:
            self.logger.info(f"Connecting to stream: {self.stream_url}")
            # In production, would establish actual stream connection
            self.stream = {'url': self.stream_url, 'protocol': self.protocol}
            self._connected = True
            self.logger.info("Stream connection established")
        except Exception as e:
            raise ConnectionError(f"Failed to connect to stream: {e}", host=self.stream_url)
    
    def disconnect(self) -> None:
        """Disconnect from stream."""
        if self.stream:
            self.logger.info("Closing stream connection")
            self.stream = None
            self._connected = False
    
    def read(self, max_messages: int = 100, **kwargs) -> Iterator[Dict[str, Any]]:
        """Read data from stream.
        
        Args:
            max_messages: Maximum messages to read
            **kwargs: Additional read parameters
            
        Yields:
            Data records as dictionaries
        """
        if not self._connected:
            raise DataIngestionError("Stream not connected", source=self.stream_url)
        
        # Mock streaming data
        for i in range(max_messages):
            yield {
                'message_id': i,
                'content': f"stream_message_{i}",
                'timestamp': '2024-01-01T00:00:00Z'
            }
    
    def get_schema(self) -> Dict[str, Any]:
        """Get stream message schema.
        
        Returns:
            Schema dictionary
        """
        return {
            'fields': [
                {'name': 'message_id', 'type': 'integer'},
                {'name': 'content', 'type': 'string'},
                {'name': 'timestamp', 'type': 'datetime'}
            ]
        }

File: domains/test/typescript/src/ingestion/sources.ts
// src/ingestion/sources.ts
/**
 * Data source implementations for the ingestion module.
 */

import { BaseComponent, generateUUID, getTimestamp } from '../core/base.js';
import { DataIngestionException, ConnectionException } from '../core/exceptions.js';
import { ComponentType, Status, DataFormat } from '../core/types.js';
import type {
  IDataSource,
  SourceType,
  SourceConfig,
  ConnectionState,
  ReadOptions,
  SourceMetadata,
  DataRecord
} from './types.js';
import type { UUID, Nullable, FilePath, URL } from '../types.js';
import * as fs from 'fs';
import * as path from 'path';
import { Readable } from 'stream';

/**
 * Abstract base data source
 */
export abstract class BaseDataSource extends BaseComponent implements IDataSource {
  public readonly type: SourceType;
  protected _connectionState: ConnectionState = ConnectionState.DISCONNECTED;
  protected config: SourceConfig;

  constructor(name: string, type: SourceType, config: SourceConfig) {
    super(name, ComponentType.INGESTER);
    this.type = type;
    this.config = config;
  }

  public get connectionState(): ConnectionState {
    return this._connectionState;
  }

  protected setConnectionState(state: ConnectionState): void {
    this._connectionState = state;
    this.emitEvent('connection:state', { state });
  }

  public abstract connect(): Promise<void>;
  public abstract disconnect(): Promise<void>;
  public abstract async *read<T = DataRecord>(options?: ReadOptions): AsyncGenerator<T, void, unknown>;
  public abstract getMetadata(): Promise<SourceMetadata>;

  public async readBatch<T = DataRecord>(size: number, options?: ReadOptions): Promise<T[]> {
    const batch: T[] = [];
    let count = 0;

    for await (const item of this.read<T>(options)) {
      batch.push(item);
      count++;
      if (count >= size) break;
    }

    return batch;
  }

  public async validate(): Promise<boolean> {
    try {
      await this.connect();
      await this.disconnect();
      return true;
    } catch {
      return false;
    }
  }

  public async initialize(): Promise<void> {
    this.logger.debug(`Initializing ${this.type} source: ${this.name}`);
  }

  public async cleanup(): Promise<void> {
    if (this.connectionState === ConnectionState.CONNECTED) {
      await this.disconnect();
    }
  }
}

/**
 * File data source
 */
export class FileSource extends BaseDataSource {
  private readonly filePath: FilePath;
  private readonly format: DataFormat;
  private fileHandle: fs.promises.FileHandle | null = null;
  private stream: fs.ReadStream | null = null;

  constructor(filePath: FilePath, format?: DataFormat, config?: Partial<SourceConfig>) {
    super(
      `file:${path.basename(filePath as string)}`,
      SourceType.FILE,
      {
        type: SourceType.FILE,
        name: path.basename(filePath as string),
        format: format || FileSource.detectFormat(filePath),
        ...config
      }
    );
    this.filePath = filePath;
    this.format = this.config.format!;
  }

  public async connect(): Promise<void> {
    try {
      this.setConnectionState(ConnectionState.CONNECTING);
      
      // Check if file exists
      await fs.promises.access(this.filePath as string, fs.constants.R_OK);
      
      // Open file handle
      this.fileHandle = await fs.promises.open(this.filePath as string, 'r');
      
      this.setConnectionState(ConnectionState.CONNECTED);
      this.logger.info(`Connected to file: ${this.filePath}`);
    } catch (error) {
      this.setConnectionState(ConnectionState.ERROR);
      throw new ConnectionException(
        `Failed to connect to file: ${this.filePath}`,
        this.filePath as string
      );
    }
  }

  public async disconnect(): Promise<void> {
    try {
      this.setConnectionState(ConnectionState.DISCONNECTING);
      
      if (this.stream) {
        this.stream.destroy();
        this.stream = null;
      }
      
      if (this.fileHandle) {
        await this.fileHandle.close();
        this.fileHandle = null;
      }
      
      this.setConnectionState(ConnectionState.DISCONNECTED);
      this.logger.info(`Disconnected from file: ${this.filePath}`);
    } catch (error) {
      this.setConnectionState(ConnectionState.ERROR);
      throw new ConnectionException(`Failed to disconnect from file: ${error}`);
    }
  }

  public async *read<T = DataRecord>(options?: ReadOptions): AsyncGenerator<T, void, unknown> {
    if (this.connectionState !== ConnectionState.CONNECTED) {
      await this.connect();
    }

    // Create read stream
    this.stream = fs.createReadStream(this.filePath as string, {
      encoding: 'utf8',
      highWaterMark: 64 * 1024 // 64KB chunks
    });

    // Read file line by line for text formats
    if (this.format === DataFormat.CSV || this.format === DataFormat.TEXT) {
      let buffer = '';
      
      for await (const chunk of this.stream) {
        buffer += chunk;
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';
        
        for (const line of lines) {
          if (line.trim()) {
            // Parse based on format
            const record = this.parseLine(line);
            if (options?.filter && !this.matchesFilter(record, options.filter)) {
              continue;
            }
            yield record as T;
          }
        }
      }
      
      // Process remaining buffer
      if (buffer.trim()) {
        yield this.parseLine(buffer) as T;
      }
    } else {
      // For other formats, read entire file
      const content = await fs.promises.readFile(this.filePath as string, 'utf8');
      const data = JSON.parse(content);
      
      if (Array.isArray(data)) {
        for (const item of data) {
          yield item as T;
        }
      } else {
        yield data as T;
      }
    }
  }

  public async getMetadata(): Promise<SourceMetadata> {
    const stats = await fs.promises.stat(this.filePath as string);
    
    return {
      id: generateUUID(),
      name: this.name,
      sourceType: this.type,
      format: this.format,
      size: stats.size,
      lastModified: stats.mtimeMs as any,
      createdAt: stats.birthtimeMs as any,
      updatedAt: stats.mtimeMs as any,
      version: '1.0.0',
      encoding: 'utf8'
    };
  }

  private parseLine(line: string): DataRecord {
    if (this.format === DataFormat.CSV) {
      // Simple CSV parsing (production would use proper parser)
      const values = line.split(',').map(v => v.trim());
      return values.reduce((acc, val, idx) => {
        acc[`field_${idx}`] = val;
        return acc;
      }, {} as DataRecord);
    }
    
    return { line };
  }

  private matchesFilter(record: DataRecord, filter: any): boolean {
    // Simple filter implementation
    return true;
  }

  private static detectFormat(filePath: FilePath): DataFormat {
    const ext = path.extname(filePath as string).toLowerCase();
    const formatMap: Record<string, DataFormat> = {
      '.json': DataFormat.JSON,
      '.csv': DataFormat.CSV,
      '.xml': DataFormat.XML,
      '.txt': DataFormat.TEXT,
      '.parquet': DataFormat.PARQUET,
      '.xlsx': DataFormat.EXCEL,
      '.xls': DataFormat.EXCEL
    };
    
    return formatMap[ext] || DataFormat.TEXT;
  }
}

/**
 * Database data source
 */
export class DatabaseSource extends BaseDataSource {
  private connection: any = null;
  private readonly connectionString: string;
  private readonly query?: string;
  private readonly table?: string;

  constructor(
    connectionString: string,
    query?: string,
    table?: string,
    config?: Partial<SourceConfig>
  ) {
    super(
      `db:${table || 'query'}`,
      SourceType.DATABASE,
      {
        type: SourceType.DATABASE,
        name: table || 'query',
        ...config
      }
    );
    this.connectionString = connectionString;
    this.query = query;
    this.table = table;
  }

  public async connect(): Promise<void> {
    try {
      this.setConnectionState(ConnectionState.CONNECTING);
      
      // Mock database connection
      this.connection = {
        connected: true,
        connectionString: this.connectionString
      };
      
      this.setConnectionState(ConnectionState.CONNECTED);
      this.logger.info('Connected to database');
    } catch (error) {
      this.setConnectionState(ConnectionState.ERROR);
      throw new ConnectionException('Failed to connect to database');
    }
  }

  public async disconnect(): Promise<void> {
    try {
      this.setConnectionState(ConnectionState.DISCONNECTING);
      
      if (this.connection) {
        this.connection.connected = false;
        this.connection = null;
      }
      
      this.setConnectionState(ConnectionState.DISCONNECTED);
      this.logger.info('Disconnected from database');
    } catch (error) {
      this.setConnectionState(ConnectionState.ERROR);
      throw new ConnectionException('Failed to disconnect from database');
    }
  }

  public async *read<T = DataRecord>(options?: ReadOptions): AsyncGenerator<T, void, unknown> {
    if (this.connectionState !== ConnectionState.CONNECTED) {
      await this.connect();
    }

    // Mock database query
    const mockData = [
      { id: 1, name: 'Record 1', value: 100 },
      { id: 2, name: 'Record 2', value: 200 },
      { id: 3, name: 'Record 3', value: 300 }
    ];

    for (const record of mockData) {
      yield record as T;
    }
  }

  public async getMetadata(): Promise<SourceMetadata> {
    return {
      id: generateUUID(),
      name: this.name,
      sourceType: this.type,
      recordCount: 3,
      createdAt: getTimestamp(),
      updatedAt: getTimestamp(),
      version: '1.0.0'
    };
  }
}

/**
 * API data source
 */
export class APISource extends BaseDataSource {
  private readonly endpoint: URL;
  private readonly method: string;
  private readonly headers: Record<string, string>;
  private session: any = null;

  constructor(
    endpoint: URL,
    method = 'GET',
    headers?: Record<string, string>,
    config?: Partial<SourceConfig>
  ) {
    super(
      `api:${endpoint}`,
      SourceType.API,
      {
        type: SourceType.API,
        name: endpoint as string,
        ...config
      }
    );
    this.endpoint = endpoint;
    this.method = method;
    this.headers = headers || {};
  }

  public async connect(): Promise<void> {
    try {
      this.setConnectionState(ConnectionState.CONNECTING);
      
      // Add authentication headers
      if (this.config.credentials) {
        if (this.config.credentials.apiKey) {
          this.headers['X-API-Key'] = this.config.credentials.apiKey;
        } else if (this.config.credentials.token) {
          this.headers['Authorization'] = `Bearer ${this.config.credentials.token}`;
        }
      }
      
      this.session = { headers: this.headers };
      this.setConnectionState(ConnectionState.CONNECTED);
      this.logger.info(`Connected to API: ${this.endpoint}`);
    } catch (error) {
      this.setConnectionState(ConnectionState.ERROR);
      throw new ConnectionException(`Failed to connect to API: ${this.endpoint}`);
    }
  }

  public async disconnect(): Promise<void> {
    this.setConnectionState(ConnectionState.DISCONNECTING);
    this.session = null;
    this.setConnectionState(ConnectionState.DISCONNECTED);
    this.logger.info(`Disconnected from API: ${this.endpoint}`);
  }

  public async *read<T = DataRecord>(options?: ReadOptions): AsyncGenerator<T, void, unknown> {
    if (this.connectionState !== ConnectionState.CONNECTED) {
      await this.connect();
    }

    // Mock API response
    const mockResponse = {
      data: [
        { id: 1, status: 'active', created: '2024-01-01' },
        { id: 2, status: 'pending', created: '2024-01-02' },
        { id: 3, status: 'active', created: '2024-01-03' }
      ]
    };

    for (const item of mockResponse.data) {
      yield item as T;
    }
  }

  public async getMetadata(): Promise<SourceMetadata> {
    return {
      id: generateUUID(),
      name: this.name,
      sourceType: this.type,
      format: DataFormat.JSON,
      createdAt: getTimestamp(),
      updatedAt: getTimestamp(),
      version: '1.0.0'
    };
  }
}

/**
 * Memory data source
 */
export class MemorySource extends BaseDataSource {
  private data: DataRecord[];

  constructor(data: DataRecord[], name = 'memory', config?: Partial<SourceConfig>) {
    super(
      name,
      SourceType.MEMORY,
      {
        type: SourceType.MEMORY,
        name,
        ...config
      }
    );
    this.data = data;
  }

  public async connect(): Promise<void> {
    this.setConnectionState(ConnectionState.CONNECTED);
    this.logger.info('Connected to memory source');
  }

  public async disconnect(): Promise<void> {
    this.setConnectionState(ConnectionState.DISCONNECTED);
    this.logger.info('Disconnected from memory source');
  }

  public async *read<T = DataRecord>(options?: ReadOptions): AsyncGenerator<T, void, unknown> {
    let items = [...this.data];
    
    // Apply offset
    if (options?.offset) {
      items = items.slice(options.offset);
    }
    
    // Apply limit
    if (options?.limit) {
      items = items.slice(0, options.limit);
    }
    
    for (const item of items) {
      yield item as T;
    }
  }

  public async getMetadata(): Promise<SourceMetadata> {
    return {
      id: generateUUID(),
      name: this.name,
      sourceType: this.type,
      recordCount: this.data.length,
      size: JSON.stringify(this.data).length,
      createdAt: getTimestamp(),
      updatedAt: getTimestamp(),
      version: '1.0.0'
    };
  }
}